### ISSUE TITLE
 "UnturnedZombieDyingEvent" sometimes does not trigger as expected.

### ISSUE THREAD

**Issue by y1meng** [openmod/openmod#827]
If the value of "DamageAmount" is modified in the "UnturnedZombieDamagingEvent" event and the value exceeds the zombie's health, an issue occurs.

**Comment by y1meng** [comment]
`
        private void DamageTool_damageZombieRequested(ref DamageZombieParameters parameters, ref bool shouldAllow)
        {
            var zombie = new UnturnedZombie(parameters.zombie);
            var player = GetUnturnedPlayer(parameters.instigator as Player);

            //from method damagetool.damageZombie
            if (parameters.applyGlobalArmorMultiplier)
            {
                parameters.times *= parameters.limb switch
                {
                    ELimb.SKULL => Provider.modeConfigData.Zombies.Armor_Multiplier,
                    _ => Provider.modeConfigData.Zombies.NonHeadshot_Armor_Multiplier,
                };
            }

            var damageAmount = (ushort)Mathf.Min(ushort.MaxValue, Mathf.FloorToInt(parameters.damage * parameters.times));
            var beaconDamageDecreasePercent = 1f;

            //from method zombie.askdamage
            var region = ZombieManager.regions[zombie.Zombie.bound];
            if (region.hasBeacon)
            {
                beaconDamageDecreasePercent = Mathf.Max(1f, BeaconManager.checkBeacon(region.nav).initialParticipants) * 1.5f;
                damageAmount = MathfEx.CeilToUShort(damageAmount / beaconDamageDecreasePercent);
            }


            var @damagingEvent = new UnturnedZombieDamagingEvent(
                zombie, player, damageAmount, parameters.direction, parameters.ragdollEffect, parameters.zombieStunOverride, parameters.limb);
            @damagingEvent.IsCancelled = !shouldAllow;
            Emit(@damagingEvent);

            var @event = @damagingEvent;

            if (@damagingEvent.DamageAmount >= zombie.Health) {
                //
                var @dyingEvent = new UnturnedZombieDyingEvent(
                    zombie, player, @damagingEvent.DamageAmount, @damagingEvent.Ragdoll, @damagingEvent.RagdollEffect, @damagingEvent.StunOverride, @damagingEvent.Limb);
                @dyingEvent.IsCancelled = !shouldAllow;
                Emit(@dyingEvent);

                @event = @dyingEvent;
            }


            // remove damage reduction due to beacon
            var newDamageAmount = @event.DamageAmount * beaconDamageDecreasePercent;

            parameters.damage = newDamageAmount;
            parameters.times = 1;
            parameters.direction = @event.Ragdoll;
            parameters.ragdollEffect = @event.RagdollEffect;
            parameters.instigator = player != null ? @event.Instigator?.Player : parameters.instigator;
            parameters.zombieStunOverride = @event.StunOverride;
            parameters.limb = @event.Limb;
            shouldAllow = !@event.IsCancelled;
        }
`
I attempted to make such changes to the event handling source code but found that it causes the "UnturnedZombieDamagingEvent" to trigger twice. Since "UnturnedZombieDyingEvent" inherits from "UnturnedZombieDamagingEvent," I realized that a separate event is needed to properly fix this issue. If the developers of OpenMod agree with my assessment, I would be happy to initiate a pull request.

**Comment by rube200** [comment]
What are you attempting to do? Every method related to damage/dying works in this way, Player, Animals, Barricades ...

I dont think this change is needed for now, can you explain you case better?

**Comment by y1meng** [comment]
> What are you attempting to do? Every method related to damage/dying works in this way, Player, Animals, Barricades ...
> 
> I dont think this change is needed for now, can you explain you case better?

My issue is that when I enhance the damage dealt by the player to the zombie while handling the "UnturnedZombieDamagingEvent" in my plugin, it causes the "UnturnedZombieDyingEvent" to not trigger. For example, if the zombie's health is 100 and the player's original damage is 50, I modify it to 100 in the "UnturnedZombieDamagingEvent" event, which causes this issue. My understanding is that the "DamageAmount" property in "UnturnedZombieDamagingEvent" is supposed to be freely modifiable by plugin developers. Does this explanation make sense to you?

**Comment by rube200** [comment]
> My issue is that when I enhance the damage dealt by the player to the zombie while handling the "UnturnedZombieDamagingEvent" in my plugin, it causes the "UnturnedZombieDyingEvent" to not trigger

I can understand that but for now every damage/dying event works in this way, we could perhaps change this but i dont think very necessary.

Plugins could also subs to `UnturnedZombieDamagingEvent` instead of `UnturnedZombieDyingEvent` and check if the damage is enought to kill, that is just a single line check, in other hand to change this behaviour we would need to duplicate alot of code on om side just for those methods

@openmod/openmod-maintainers Should we make the change? What do you think?

**Comment by y1meng** [comment]
> > My issue is that when I enhance the damage dealt by the player to the zombie while handling the "UnturnedZombieDamagingEvent" in my plugin, it causes the "UnturnedZombieDyingEvent" to not trigger
> 
> I can understand that but for now every damage/dying event works in this way, we could perhaps change this but i dont think very necessary.
> 
> Plugins could also subs to `UnturnedZombieDamagingEvent` instead of `UnturnedZombieDyingEvent` and check if the damage is enought to kill, that is just a single line check, in other hand to change this behaviour we would need to duplicate alot of code on om side just for those methods
> 
> @openmod/openmod-maintainers Should we make the change? What do you think?

Yes, I have now changed it so that the plugin itself determines if the zombie will die, no longer using the "UnturnedZombieDyingEvent." However, this can indeed cause confusion for plugin developers. You might want to evaluate this and consider adding a separate "PlayerKillZombieEvent" to avoid too many changes(lol.

**Comment by liebesleid-eth** [comment]
I tried using both events just to know if a player killed a zombie and it would be wonky, it'd trigger the event without actually killing the zombie, then again once you hit the zombie once more.

Not sure if I was doing something dumb, but switching to unturned's stat increase event seemed to work fine (though no longer was able to get zombie specialty or don't know how :<)
